libs/libarchfpga/src/physical_types_util.cpp

```diff
@@ -35,35 +36,102 @@ struct t_pin_inst_port {
 
 /******************** Subroutine declarations and definition ****************************/
 
-static std::tuple<int, int, int> get_pin_index_for_inst(t_physical_tile_type_ptr type, int pin_index);
-static t_pin_inst_port block_type_pin_index_to_pin_inst(t_physical_tile_type_ptr type, int pin_index);
+static std::tuple<int, int, int, int> get_pin_index_for_inst(t_physical_tile_type_ptr type, int pin_index, bool is_flat);
 
-static std::tuple<int, int, int> get_pin_index_for_inst(t_physical_tile_type_ptr type, int pin_index) {
-    VTR_ASSERT(pin_index < type->num_pins);
+// #TODO: remove the default value for is_flat
+static t_pin_inst_port block_type_pin_index_to_pin_inst(t_physical_tile_type_ptr type, int pin_index, bool is_flat = false);
 
-    int total_pin_counts = 0;
-    int pin_offset = 0;
-    for (auto& sub_tile : type->sub_tiles) {
-        total_pin_counts += sub_tile.num_phy_pins;
+static int get_sub_tile_num_internal_classes(const t_sub_tile* sub_tile);
```

Comment:
We need to add comments on the new function explaning what do these function do and how to use them: what are the inputs and what are the outputs.